File: wordLadder.java
// Time Complexity :O(n*m*m*26) where n is the number of words in the wordList and m is the length of each word
// Space Complexity : O(n) where n is the number of words in the wordList
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
/*
 * 1. Add all the words in the wordList to a set.
 * 2. Add the beginWord to the queue to start BFS. 
 * 3. For each word in the queue, replace each character with a-z and check if the new word is in the set. If it is, add it to the queue and remove it from the set.
 * 4. If the new word is the endWord, return the depth+1.
 */

import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        HashSet<String> set = new HashSet<>();
        for(String word : wordList){
            set.add(word);
        }

        Queue<String> queue = new LinkedList<>();
        queue.offer(beginWord);
        int depth = 0;

        while(!queue.isEmpty()){
            int size = queue.size();
            depth++;

            for(int i=0; i<size; i++){
                String currWord = queue.poll();
                char[] currWordCh = currWord.toCharArray();
                for(int j=0; j<currWordCh.length; j++){
                    char prevch = currWordCh[j];
                    for(int k=0; k<26; k++){
                        char replaceChar = (char)('a' + k);
                        if(replaceChar == prevch)
                            continue;
                        
                        currWordCh[j] = replaceChar;
                        String newWord = new String(currWordCh);
                        if(!set.contains(newWord))
                            continue;
                        if(newWord.equals(endWord))
                            return depth+1;

                        set.remove(newWord);
                        queue.offer(newWord);
                    }

                    currWordCh[j] = prevch;
                }
            }
        }

        return 0;
    }
}

File: WordLadder.cs
using System;
using System.Collections.Generic;

namespace Algorithms
{
    /// Time Complexity : O(M2×N), where MM is the length of each word and NN is the total number of words in the input word list.
    // Space Complexity :O(M2×N), where MM is the length of each word and NN is the total number of words in the input word list.
    // Did this code successfully run on Leetcode :Yes
    // Any problem you faced while coding this :  No 
    public class WordLadder
    {
        public int LadderLength(string beginWord, string endWord, IList<string> wordList)
        {
            HashSet<string> set = new HashSet<string>(wordList);
            if (beginWord == endWord)
                return 0;
            if (!set.Contains(endWord))
                return 0;

            int minNumberOfTransactions = 0, wordlength = beginWord.Length;

            Queue<string> q = new Queue<string>();
            q.Enqueue(beginWord);

            while (q.Count != 0)
            {
                minNumberOfTransactions++;
                int size = q.Count;
                for (int i = 0; i < size; ++i)
                {
                    char[] currWord = q.Dequeue().ToCharArray();

                    for (int pos = 0; pos < wordlength; ++pos)
                    {
                        char orig_char = currWord[pos];
                        for (char c = 'a'; c <= 'z'; ++c)
                        {
                            currWord[pos] = c;

                            if (String.Join("", currWord).Equals(endWord))
                                return minNumberOfTransactions + 1;

                            if (!set.Contains(String.Join("", currWord)))
                                continue;
                            set.Remove(String.Join("", currWord));

                            q.Enqueue(String.Join("", currWord));
                        }
                        currWord[pos] = orig_char;
                    }
                }
            }
            return 0;

        }
    }
}
